File: packages/node/src/tokenize.ts
// Step 4 of the parsing procedure (docs/nomenclature.md section 12).

export interface Token {
  /** Lowercase, accents stripped (ñ kept). Used for matching. */
  norm: string;
  /** The same characters as they appear in the input. */
  orig: string;
  isNum: boolean;
  /** True when no separator stands between this token and the previous one: the "A" in "45A". */
  attached: boolean;
  start: number;
  end: number;
}

type CharClass = "letter" | "digit" | "separator";

/** Normalizes one character for matching. */
export function normalizeChar(char: string): string {
  if (char === "ñ" || char === "Ñ") return "ñ";
  if (char === "º" || char === "°") return "o";
  return char.normalize("NFD").replace(/\p{M}/gu, "").toLowerCase();
}

function classify(norm: string): CharClass {
  if (/^[a-zñ]$/.test(norm)) return "letter";
  if (/^[0-9]$/.test(norm)) return "digit";
  return "separator";
}

/**
 * Splits on spaces, "#", hyphens, dashes, periods and any other punctuation,
 * and on letter-digit boundaries: "KR45" -> "KR", "45" and "45A" -> "45", "A".
 */
export function tokenize(text: string): Token[] {
  const tokens: Token[] = [];
  let current: { cls: CharClass; start: number; norm: string } | null = null;
  let lastEnd = -1;

  const flush = (end: number) => {
    if (!current) return;
    tokens.push({
      norm: current.norm,
      orig: text.slice(current.start, end),
      isNum: current.cls === "digit",
      attached: tokens.length > 0 && current.start === lastEnd,
      start: current.start,
      end,
    });
    lastEnd = end;
    current = null;
  };

  const chars = Array.from(text);
  let offset = 0;
  for (const char of chars) {
    const norm = normalizeChar(char);
    const cls = classify(norm);
    if (cls === "separator") {
      flush(offset);
    } else if (current && current.cls === cls) {
      current.norm += norm;
    } else {
      flush(offset);
      current = { cls, start: offset, norm };
    }
    offset += char.length;
  }
  flush(offset);
  return tokens;
}

File: packages/python/src/geographicol/tokenize.py
"""Step 4 of the parsing procedure (docs/nomenclature.md section 12)."""

from __future__ import annotations

import unicodedata
from dataclasses import dataclass
from typing import Literal

CharClass = Literal["letter", "digit", "separator"]

_LETTERS = set("abcdefghijklmnopqrstuvwxyzñ")
_DIGITS = set("0123456789")


@dataclass(frozen=True)
class Token:
    norm: str
    """Lowercase, accents stripped (ñ kept). Used for matching."""
    orig: str
    """The same characters as they appear in the input."""
    is_num: bool
    attached: bool
    """True when no separator stands between this token and the previous one: the "A" in "45A"."""
    start: int
    end: int


def normalize_char(char: str) -> str:
    """Normalizes one character for matching."""
    if char in ("ñ", "Ñ"):
        return "ñ"
    if char in ("º", "°"):
        return "o"
    decomposed = unicodedata.normalize("NFD", char)
    # Drop every combining mark (Unicode category M), matching /\p{M}/u in the Node library.
    return "".join(c for c in decomposed if not unicodedata.category(c).startswith("M")).lower()


def _classify(norm: str) -> CharClass:
    if norm in _LETTERS:
        return "letter"
    if norm in _DIGITS:
        return "digit"
    return "separator"


def tokenize(text: str) -> list[Token]:
    """Splits on spaces, "#", hyphens, dashes, periods and any other punctuation,
    and on letter-digit boundaries: "KR45" -> "KR", "45" and "45A" -> "45", "A".
    """
    tokens: list[Token] = []
    cls: CharClass | None = None
    start = 0
    norm = ""
    last_end = -1

    def flush(end: int) -> None:
        nonlocal cls, last_end
        if cls is None:
            return
        tokens.append(
            Token(
                norm=norm,
                orig=text[start:end],
                is_num=cls == "digit",
                attached=bool(tokens) and start == last_end,
                start=start,
                end=end,
            )
        )
        last_end = end
        cls = None

    for offset, char in enumerate(text):
        char_norm = normalize_char(char)
        char_cls = _classify(char_norm)
        if char_cls == "separator":
            flush(offset)
        elif cls == char_cls:
            norm += char_norm
        else:
            flush(offset)
            cls, start, norm = char_cls, offset, char_norm
    flush(len(text))
    return tokens
